pkg/keys/hashicorp_vault.go

```diff
@@ -54,20 +54,97 @@ func NewHashiCorpVault(ctx context.Context) (KeyManager, error) {
 	return sm, nil
 }
 
+type HCValueKeyID struct {
+	Name    string
+	Version string
+}
+
+func NewHCValueKeyID(keyID string) (*HCValueKeyID, error) {
+	parts := strings.SplitN(keyID, "@", 2)
+	switch len(parts) {
+	case 0, 1:
+		return nil, fmt.Errorf("missing version in: %v", keyID)
+	default:
+		return &HCValueKeyID{Name: parts[0], Version: parts[1]}, nil
+	}
+}
+
 // NewSigner creates a new signer that uses a key in HashiCorp Vault's transit
 // backend. The keyID is in the format:
 //
 //     name@version
 //
 // Both name and version are required.
 func (v *HashiCorpVault) NewSigner(ctx context.Context, keyID string) (crypto.Signer, error) {
-	parts := strings.SplitN(keyID, "@", 2)
-	switch len(parts) {
-	case 0, 1:
-		return nil, fmt.Errorf("missing version in: %v", keyID)
-	default:
-		return NewHashiCorpVaultSigner(ctx, v.client, parts[0], parts[1])
+	hvcKey, err := NewHCValueKeyID(keyID)
+	if err != nil {
+		return nil, err
+	}
+	return NewHashiCorpVaultSigner(ctx, v.client, hvcKey.Name, hvcKey.Version)
+}
+
+func (v *HashiCorpVault) Encrypt(ctx context.Context, keyID string, plaintext []byte, aad string) ([]byte, error) {
+	kid, err := NewHCValueKeyID(keyID)
+	if err != nil {
+		return nil, err
+	}
+	pth := fmt.Sprintf("transit/encrypt/%s", kid.Name)
+	params := map[string]interface{}{
+		"plaintext": base64.StdEncoding.EncodeToString(plaintext),
+		"context":   base64.StdEncoding.EncodeToString([]byte(aad)),
+		"type":      "aes256-gcm96",
+	}
+
+	result, err := v.client.Logical().Write(pth, params)
+	if err != nil {
+		return nil, fmt.Errorf("unable to encrypt: %w", err)
 	}
+
+	ciphertext, ok := result.Data["ciphertext"]
+	if !ok {
+		return nil, fmt.Errorf("encryption returned no ciphertext")
+	}
+
+	parts := strings.Split(ciphertext.(string), ":")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("invalid encryption result, incorrect format")
+	}
+
+	cipherbytes, err := base64util.DecodeString(parts[2])
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode ciphertext: %w", err)
+	}
+
+	return cipherbytes, nil
+}
+
+func (v *HashiCorpVault) Decrypt(ctx context.Context, keyID string, ciphertext []byte, aad string) ([]byte, error) {
+	kid, err := NewHCValueKeyID(keyID)
+	if err != nil {
+		return nil, err
+	}
+	pth := fmt.Sprintf("transit/decrypt/%s", kid.Name)
+	params := map[string]interface{}{
+		"ciphertext": fmt.Sprintf("vault:v1:%s", base64.StdEncoding.EncodeToString(ciphertext)),
```

Comment:
It won't always be v1 - that's why we should store the full blob that's returned.